jetty-server/src/main/java/org/eclipse/jetty/server/PlainOrSslConnectionFactory.java

```diff
@@ -0,0 +1,192 @@
+        int tlsFrameType = buffer.get(0) & 0xFF;
+        int tlsMajorVersion = buffer.get(1) & 0xFF;
+        return (tlsFrameType == TLS_HANDSHAKE_FRAME_TYPE || tlsFrameType == TLS_ALERT_FRAME_TYPE) && tlsMajorVersion == TLS_MAJOR_VERSION;
+    }
+
+    /**
+     * <p>Callback method invoked when {@code plainProtocol} is {@code null}
+     * and the first bytes are not TLS.</p>
+     * <p>This typically happens when a client is trying to connect to a TLS
+     * port using the {@code http} scheme (and not the {@code https} scheme).</p>
+     * <p>This method may be overridden to write back a minimal response such as:</p>
+     * <pre>
+     * HTTP/1.1 400 Bad Request
+     * Content-Length: 35
+     * Content-Type: text/plain; charset=UTF8
+     * Connection: close
+     *
+     * Plain HTTP request sent to TLS port
+     * </pre>
+     *
+     * @param buffer   The buffer with the first bytes of the connection
+     * @param endPoint The connection EndPoint object
+     * @see #seemsTLS(ByteBuffer)
+     */
+    protected void unknownProtocol(ByteBuffer buffer, EndPoint endPoint)
+    {
+        endPoint.close();
+    }
+
+    private class PlainOrSslConnection extends AbstractConnection implements Connection.UpgradeFrom
+    {
+        private final Connector connector;
+        private final ByteBuffer buffer;
+
+        public PlainOrSslConnection(EndPoint endPoint, Connector connector)
+        {
+            super(endPoint, connector.getExecutor());
+            this.connector = connector;
+            this.buffer = BufferUtil.allocateDirect(1536);
+        }
+
+        @Override
+        public void onOpen()
+        {
+            super.onOpen();
+            fillInterested();
+        }
+
+        @Override
+        public void onFillable()
+        {
+            try
+            {
+                int filled = getEndPoint().fill(buffer);
+                if (filled > 0)
+                {
+                    upgrade(buffer);
+                }
+                else if (filled == 0)
+                {
+                    fillInterested();
+                }
+                else
+                {
+                    close();
+                }
+            }
+            catch (IOException x)
+            {
+                LOG.warn(x);
+                close();
+            }
+        }
+
+        @Override
+        public ByteBuffer onUpgradeFrom()
+        {
+            return buffer;
+        }
+
+        private void upgrade(ByteBuffer buffer)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Read {}", BufferUtil.toDetailString(buffer));
+
+            EndPoint endPoint = getEndPoint();
+            if (seemsTLS(buffer))
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Detected TLS bytes, upgrading to {}", sslConnectionFactory);
+                endPoint.upgrade(sslConnectionFactory.newConnection(connector, endPoint));
+            }
+            else
```

Comment:
Should we be able to optionally check for ascii characters if we know the other protocol is a plain text one?

Reply:
We don't know what the other protocol is, so it may not be ASCII.
I would just test for the first TLS byte, and if not, just do what the configuration tells me to.
If the user has configured `plainProtocol` to be `fcgi` so be it.

Reply:
Then we should definitely test as many bytes as possible and `plainProtocol` is the wrong name - perhaps `otherProtocol`